server/linkpreview.go

```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"encoding/json"
+	"golang.org/x/net/html"
+	"net/http"
+	"strings"
+	"time"
+)
+
+type LinkPreview struct {
+	Title       string `json:"title"`
+	Description string `json:"description"`
+	Image       string `json:"image"`
+	URL         string `json:"url"`
+}
+
+// PreviewLink handles the HTTP request, fetches the URL, and returns the link preview
+func PreviewLink(w http.ResponseWriter, r *http.Request) {
+	url := r.URL.Query().Get("url")
+	if url == "" {
+		http.Error(w, "Missing 'url' query parameter", http.StatusBadRequest)
+		return
+	}
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		http.Error(w, "Non-OK HTTP status", http.StatusInternalServerError)
```

Comment:
If response code is 20x, then you should not return an error here. Try parsing the response anyway. If Content-Length is 0, return a valid response with empty content.
For the rest of responses shouldn't be 500. Probably 502.